Removing duplicate messages in PO file parsers

phlax opened the pull request:
PO files can't cope with duplicate messages, so they are removed when parsing
and a warning is logged

Fixes translate/pootle#3564

unho commented on translate/storage/pypo.py:
```diff
@@ -759,16 +761,16 @@ class pofile(pocommon.pofile):
+        # clear units to get rid of automatically generated headers before parsing
+        self.units = []
+        poparser.parse_units(poparser.ParseState(input, pounit), self)
+        self.removeduplicates(duplicatestyle="discard")
```
Please add a commit that cleans that superfluous `if` so we can separate the cleanup for the actual changes.

unho commented on translate/storage/pypo.py:
```diff
@@ -804,6 +806,11 @@ def addcomment(thepo):
+                    filename = ''
+                    if self.filename:
+                        filename = " in '%s'" % self.filename
+                    logger.warn("Duplicate message ignored%s: '%s'" % (filename, id))
```
The following looks less convoluted to me:

```python
                else:
                    if self.filename:
                        logger.warn("Duplicate message ignored in '%s': '%s'" %
                                    (self.filename, id))

                    logger.warn("Duplicate message ignored: '%s'" % id)
```

unho replied:
Shouldn't be `logger.warning()` instead?

unho commented on translate/storage/pypo.py:
```diff
@@ -35,6 +36,7 @@
 from translate.storage import base, pocommon, poparser
 from translate.storage.pocommon import encodingToUse
 
+logger = logging.getLogger(__name__)
```
Do add an extra blank line before this.

unho commented on translate/storage/fpo.py:
```diff
@@ -540,6 +543,11 @@ def addcomment(thepo):
+                    filename = ''
+                    if self.filename:
+                        filename = " in '%s'" % self.filename
+                    logger.warn("Duplicate message ignored%s: '%s'" % (filename, id))
```
See comment below about less convoluted syntax and `logger.warning()`.

unho commented on translate/storage/cpo.py:
```diff
@@ -666,6 +666,11 @@ def addcomment(thepo):
+                    filename = ''
+                    if self.filename:
+                        filename = " in '%s'" % self.filename
+                    logger.warn("Duplicate message ignored%s: '%s'" % (filename, id))
```
See comment below about using a simpler syntax and using `logger.warning()`.

unho commented:
I am concerned about that this breaks several tests. Have you considered altering `parse()` so you can control whether duplicates are removed or not?

phlax commented:
@unho yep i was working on that earlier

I should be ready to push shortly

phlax commented:
The thing with this though is that it changes the default behaviour.

I think this is correct because it should run removedupes on parse and return valid po by default

phlax commented:
@unho Ive tidied this up a bit, fixed the tests and added a couple of tests

I have made the default for removeduplicates to discard - we could set this back to merge and still have it run removeduplicates on parse.

unho commented on translate/convert/test_po2ts.py:
```diff
@@ -77,9 +77,7 @@ def test_duplicates(self):
         tsfile = self.po2ts(posource)
-        print(tsfile)
-        assert tsfile.find("English") != tsfile.rfind("English")
-
```
Please do not remove this blank line. Please do follow PEP8 here.

phlax replied:
Apologies - removing

unho commented on translate/storage/fpo.py:
```diff
@@ -31,12 +31,14 @@
 from translate.storage import base, cpo, pocommon, poparser
 from translate.storage.pocommon import encodingToUse
 
+logger = logging.getLogger(__name__)
```
Please do add an extra blank line between this and the imports.

unho commented on translate/storage/test_po.py:
```diff
@@ -588,12 +587,11 @@ def test_multiline_obsolete(self):
         assert len(pofile.units) == 1
         assert pofile.units[0].getlocations() == ["source1", "source2"]
-        print(pofile)
+
```
Please remove this blank line. PEP8 recommends single blank line between methods.

unho commented on translate/storage/pypo.py:
```diff
@@ -804,11 +810,17 @@ def addcomment(thepo):
                         markedpos.append(thepo)
                     thepo.msgctxt.append('"%s"' % escapeforpo(" ".join(thepo.getlocations())))
                     uniqueunits.append(thepo)
+                else:
```
IMHO we should check here for `discard`.

phlax replied:
the reason i didnt was that anything you put here would discard it anyway just not trigger the warning

unho replied:
ok

unho commented on translate/storage/pypo.py:
```diff
@@ -757,20 +760,23 @@ class pofile(pocommon.pofile):
+        if duplicatestyle != "allow":
+            self.removeduplicates(duplicatestyle=duplicatestyle)
+
+    def removeduplicates(self, duplicatestyle="discard"):
```
IMHO we should keep the previous default style since that probably will require less changes to TTK.

phlax replied:
k - i reset teh default to merge for removeduplicates

this broke 0 tests - which kinda suggests removeduplicates' default needs to be tested...

unho commented on translate/storage/pypo.py:
```diff
@@ -757,20 +760,23 @@ class pofile(pocommon.pofile):
+        poparser.parse_units(poparser.ParseState(input, pounit), self)
+
+        # duplicates are now removed by default unless duplicatestyle=allow
+        if duplicatestyle != "allow":
```
Either check against `None` here or use default `allow` in the method parameter above.

phlax replied:
im not really following

we cant have duplicatestyle="allow" as the default in parse or it wouldnt fix the problem - it would revert to old behaviour

unho replied:
Just noticed what a stupid thing is to have `def parse(self, input, duplicatestyle="allow"):` since having mutable objects as default values is highly discouraged. So it should be instead:

```python
        if duplicatestyle is not None and duplicatestyle != "allow":
```


@phlax What I was expecting of this PR is to keep previous behavior by default, but allowing to discard duplicates in case it was explicitly specified when calling `parse()` (necessary for Pootle). I was hoping that this would be a non-invasive change that didn't require to touch existing tests at all, despite it would be necessary to add new tests to test new behavior. Have you tried that approach?

unho replied:
> Just noticed what a stupid thing is to have `def parse(self, input, duplicatestyle="allow"):` since having mutable objects as default values is highly discouraged. So it should be instead:

>```python
        if duplicatestyle is not None and duplicatestyle != "allow":
```

I actually think that this is probably better just at the beginning of `removeduplicates()`.

phlax replied:
```
 if duplicatestyle is not None and duplicatestyle != "allow":
```
looks good for removeduplicates

strings are *immutable*

phlax replied:
@unho 

>@phlax What I was expecting of this PR is to keep previous behavior by default, but allowing > to discard duplicates in case it was explicitly specified when calling parse() (necessary for 
> Pootle)

we can do that but that would be more invasive for because when get_class(file)() gets called it would need to have the duplicatestyle param set - therefore all of the storagetypes would need to implement it or we would need to hardcode against po files

unho commented:
@phlax Can you please rebase and merge the "superfluous if True" commit directly to `master`? Thanks in advance.

phlax commented:
@unho i dont have push power on upstream master

do you want a pull request?

unho commented on translate/convert/test_po2ts.py:
```diff
@@ -77,8 +77,7 @@ def test_duplicates(self):
 msgstr "b"
 '''
         tsfile = self.po2ts(posource)
-        print(tsfile)
```
What is the reason to remove this?

phlax replied:
i assumed that these were debugging

but, yep, i should have put this cleanup in a separate commit

phlax replied:
now in separate commit

unho replied:
I think they are for debugging, for when some test fails and you can simply go up and see what the prints did output before the failure. Does this provide some noticeable performance improvement? If yes please move this to a separate PR.

unho commented on translate/storage/test_cpo.py:
```diff
@@ -111,12 +109,9 @@ def test_merge_duplicates_msgctxt(self):
         pofile.removeduplicates("merge")
         assert len(pofile.units) == 2
-        print(pofile.units[0].msgidcomments)
-        print(pofile.units[1].msgidcomments)
```
Please do not remove these print calls. Perhaps they are not necessary, but unless they are strictly necessary to be removed this is a cleanup that doesn't belong to this commit and makes more difficult to review the actual changes.

phlax replied:
k, agreed - ill fix

phlax replied:
now in separate commit

unho commented on translate/storage/test_cpo.py:
```diff
@@ -139,8 +133,7 @@ def test_msgid_comment(self):
         """checks that keeping keeps blanks and doesn't add msgid_comments"""
         posource = '#: source1\nmsgid ""\nmsgstr ""\n\n#: source2\nmsgid ""\nmsgstr ""\n'
-        pofile = self.poparse(posource)
-        assert len(pofile.units) == 2
```
Is there any reason to remove these assert calls?

phlax replied:
yes - this assert is call is now made in line 138

phlax commented:
@unho ive removed the cleanup code and tidied up

unho commented on translate/storage/test_po.py:
```diff
@@ -18,7 +18,8 @@ def test_roundtrip_quoting():
         unquoted_special = pypo.unquotefrompo(quoted_special)
-        print("special: %r\nquoted: %r\nunquoted: %r\n" % (special, quoted_special, unquoted_special))
+        # print("special: %r\nquoted: %r\nunquoted: %r\n" % (
+        #     special, quoted_special, unquoted_special))
```
This is commented rather than removed.

phlax replied:
yep - but i think ill just remove it

unho commented:
@phlax In the future **please** do not bring more and more unrelated cleanup commits into a PR. Instead do create new PRs since those cleanups most likely are going to be quickly merged and they actually difficult reviewing the former PR.

unho commented on translate/storage/cpo.py:
```diff
@@ -465,7 +475,7 @@ def merge(self, otherpo, overwrite=False, comments=True, authoritative=False):
 
     def isheader(self):
-        #return self.source == u"" and self.target != u""
+        # return self.source == u"" and self.target != u""
```
It is better to remove the commented code.

unho commented on translate/storage/cpo.py:
```diff
@@ -492,8 +502,9 @@ def _domarkfuzzy(self, present=True):
-        # but a better solution would be for libgettextpo to output correct data on serialisation
+        # FIXME: libgettexpo currently does not reset other data, we probably
+        # want to do that but a better solution would be for libgettextpo to
+        # output correct data on serialisation
```
As per our styleguide http://docs.translatehouse.org/projects/translate-toolkit/en/latest/developers/styleguide.html#comments Write sentences correctly: proper capitalisation and punctuation.

phlax replied:
I didn't write this sentence!

unho commented on translate/storage/cpo.py:
```diff
@@ -687,8 +709,8 @@ def addcomment(thepo):
-            # Remove all items that are not output by msgmerge when a unit is obsolete.  This is a work
-            # around for bug in libgettextpo
+            # Remove all items that are not output by msgmerge when a unit is
+            # obsolete.  This is a work around for bug in libgettextpo
```
Same comment about comment punctuation as above.

phlax replied:
Same comment as abvoe - re *pep8*

unho commented on translate/storage/fpo.py:
```diff
@@ -32,6 +32,7 @@
 import re
 import six
 from io import BytesIO
+import logging
```
Please place this comment in the proper place. Check our styleguide http://docs.translatehouse.org/projects/translate-toolkit/en/latest/developers/styleguide.html#imports

phlax replied:
logging is in python stdlib - ie in the right place

unho commented on translate/storage/fpo.py:
```diff
@@ -107,13 +110,13 @@ def gettarget(self):
-#        assert isinstance(target, unicode)
-#        target = data.forceunicode(target)
+        #        assert isinstance(target, unicode)
+        #        target = data.forceunicode(target)
```
Better drop commented code.

phlax replied:
cool - ill remove

unho commented on translate/storage/fpo.py:
```diff
@@ -207,7 +210,7 @@ def merge(self, otherpo, overwrite=False, comments=True, authoritative=False):
 
         def mergelists(list1, list2, split=False):
-            #decode where necessary
+            # decode where necessary
```
Proper capitalization please.

unho commented on translate/storage/cpo.py:
```diff
@@ -575,7 +587,6 @@ def buildfromunit(cls, unit, encoding=None):
         elif isinstance(unit, pocommon.pounit):
             newunit = cls(unit.source, encoding)
             newunit.target = unit.target
-            #context
```
Here you are removing the comment while below it is arranged properly. I prefer it to be arranged properly since seems to be communicating something useful.

phlax replied:
I dont think it was. I think it was a bit of left over debugging

unho commented on translate/storage/fpo.py:
```diff
@@ -485,10 +488,10 @@ def _build_cpo_from_self(self):
                 self._cpo_store.addunit(cpo.pofile.UnitClass.buildfromunit(unit, self._encoding))
         if not self._cpo_store.header():
-            #only add a temporary header
+            # only add a temporary header
```
Proper capitalization please.

unho commented on translate/storage/fpo.py:
```diff
@@ -496,9 +499,9 @@ def parse(self, input):
-#                tmp_header_added = True
+            # if isinstance(input, str) and '"Content-Type: text/plain; charset=' not in input[:200]:
+            #     input = basic_header + input
+            #     tmp_header_added = True
```
Better remove commented code.

phlax replied:
k

unho commented on translate/storage/pocommon.py:
```diff
@@ -90,7 +90,10 @@ def geterrors(self):
+        """
+        Marks the unit to indicate whether it needs review. Adds an optional
+explanation as a note.
+        """
```
I think it is preferable to use:

```python
        """Marks the unit to indicate whether it needs review.

        Adds an optional explanation as a note.
        """
```

unho commented on translate/storage/pocommon.py:
```diff
@@ -196,18 +207,19 @@ def encodingToUse(encoding):
     Name = "Gettext PO file"  # pylint: disable=E0602
-    Mimetypes = ["text/x-gettext-catalog", "text/x-gettext-translation", "text/x-po", "text/x-pot"]
+    Mimetypes = ["text/x-gettext-catalog", "text/x-gettext-translation",
+                 "text/x-po", "text/x-pot"]
```
From our styleguide: For lists or tuples with many items, break immediately after the opening brace

So:

```python
    Mimetypes = [
        "text/x-gettext-catalog",
        "text/x-gettext-translation",
        "text/x-po",
        "text/x-pot",
    ]
```

phlax replied:
it really pains me to do that - it breaks pep8

claudep replied:
Absolutely not (AFAIK).

phlax replied:
My preferred style is:

```
 Mimetypes = [
       "text/x-gettext-catalog",
       "text/x-gettext-translation",
       "text/x-po",
       "text/x-pot"]
```

phlax replied:
actually your right - its just emacs wont allow it - it forces:

```
Mimetypes = [
    "text/x-gettext-catalog",
    "text/x-gettext-translation",
    "text/x-po",
     "text/x-pot",
     ]

unho commented on translate/storage/pypo.py:
```diff
@@ -28,15 +28,18 @@
 import six
 import textwrap
 from io import BytesIO
+import logging
```
Please place import in the right place according to our styleguide.

unho commented on translate/storage/pypo.py:
```diff
@@ -411,7 +424,7 @@ def merge(self, otherpo, overwrite=False, comments=True, authoritative=False):
 
         def mergelists(list1, list2, split=False):
-            #decode where necessary
+            # decode where necessary
```
Please use proper capitalization.

unho commented on translate/storage/pypo.py:
```diff
@@ -482,8 +498,10 @@ def mergelists(list1, list2, split=False):
-        #rewritten here for performance:
+        # return (self._msgidlen() == 0)
+        # and (self._msgstrlen() > 0)
+        # and (len(self.msgidcomments) == 0)
```
Remove commented code.

unho commented on translate/storage/pypo.py:
```diff
@@ -482,8 +498,10 @@ def mergelists(list1, list2, split=False):
+        # return (self._msgidlen() == 0)
+        # and (self._msgstrlen() > 0)
+        # and (len(self.msgidcomments) == 0)
+        # rewritten here for performance:
```
If the commented code is removed then this comment is useless.

unho commented on translate/storage/pypo.py:
```diff
@@ -804,24 +836,32 @@ def addcomment(thepo):
+        """
+        Convert to a string.
+        Double check that unicode is handled somehow here
+        """
```
Better:

```python
        """ Convert to a string.

        Double check that unicode is handled somehow here
        """
```

phlax replied:
I dont really agree on this one

unho commented on translate/storage/test_po.py:
```diff
@@ -435,8 +417,8 @@ def test_uniforum_po(self):
-        #assert pofile.units[0].getlocations() == "File: somefile, line: 300"
-        #assert pofile.units[1].getlocations() == "File: anotherfile, line: 200"
+        # assert pofile.units[0].getlocations() == "File: somefile, line: 300"
+        # assert pofile.units[1].getlocations() == "File: anotherfile, line: 200"
```
Note that this commented code must be kept since is placed there for future use as previous comment states.

unho commented on translate/storage/test_po.py:
```diff
@@ -537,64 +511,65 @@ def test_makeobsolete_plural(self):
+                    'msgctxt "Context"\n'
+                    'msgid "test"\nmsgstr "rest"\n')
+        poexpected = ('#~ msgctxt "Context"\n'
+                      '#~ msgid "test"\n#~ msgstr "rest"\n')
```
You seem to have split this into several lines for readability. I like that. But it seems the `msgstr` in `poexpected` is not on its own line.

unho commented on translate/storage/test_po.py:
```diff
@@ -537,64 +511,65 @@ def test_makeobsolete_plural(self):
+                    '#~ "Second old thing"\n'
+                    '#~ msgstr ""\n'
+                    '#~ "Ou ding\\n"\n'
+                    '#~ "Tweede ou ding"\n')
```
In other test files the source is put in a triple quoted string, which allows to just copy-paste the real sample because we don't need to escape. Also it is easier to read. I suggest you to rework all these `posource` to use triple quoted strings, like:

```python
        posource = """#~ msgid ""
#~ "Old thing\n"
#~ "Second old thing"
#~ msgstr ""
#~ "Ou ding\n"
#~ "Tweede ou ding"
"""
```

phlax replied:
i avoid triple quoted strings except at module level

unho commented on translate/storage/test_po.py:
```diff
@@ -749,7 +722,8 @@ def test_id(self):
 
-    def test_non_ascii_header_comments(self):
+    # method with same name below?
+    def __test_non_ascii_header_comments(self):
```
Does this test work now that its name starts with `__test` instead of `test`? Anyway I suggest renaming the method by changing what is after `test`.

phlax replied:
no it doesnt - but as it wasnt being called already i didnt want to activate it

phlax replied:
i think with this function we should either:
- remove it altogether, or
- integrate what is there into the test below

unho commented:
@phlax Will review the actual changes re duplicates when the unrelated cleanups are out of this PR.

phlax commented:
@unho good to review

unho commented on translate/storage/fpo.py:
```diff
@@ -31,6 +31,7 @@
 import copy
 import re
 import six
+import logging
```
This import must be two lines up. Imports must be alphabetically arranged. See our styleguide on imports.

unho commented on translate/storage/pypo.py:
```diff
@@ -27,6 +27,7 @@
 import re
 import six
 import textwrap
+import logging
```
This import must be put in the right place. See comment about imports being alphabetically ordered.

unho commented on translate/storage/cpo.py:
```diff
@@ -735,7 +741,7 @@ def isempty(self):
         return True
 
-    def parse(self, input):
+    def parse(self, input, duplicatestyle=None):
```
Please provide a default value. The same goes for the other two implementations of this method.

unho replied:
Forgot about this.

unho commented on translate/storage/test_cpo.py:
```diff
@@ -99,8 +99,7 @@ def test_msgidcomments(self):
-        pofile = self.poparse(posource)
-        assert len(pofile.units) == 2
+        pofile = self.poparse(posource, duplicatestyle="allow")
         pofile.removeduplicates("msgctxt")
```
I think it is better to instead do the following:

```python
        pofile = self.poparse(posource, duplicatestyle="msgctxt")
```

unho commented on translate/storage/test_cpo.py:
```diff
@@ -111,8 +110,7 @@ def test_merge_duplicates_msgctxt(self):
-        pofile = self.poparse(posource)
-        assert len(pofile.units) == 2
+        pofile = self.poparse(posource, duplicatestyle="allow")
         pofile.removeduplicates("merge")
```
I think it is better to instead do the following:

```python
        pofile = self.poparse(posource, duplicatestyle="merge")
```

unho commented on translate/storage/test_cpo.py:
```diff
@@ -124,8 +122,7 @@ def test_merge_blanks(self):
-        pofile = self.poparse(posource)
-        assert len(pofile.units) == 2
+        pofile = self.poparse(posource, duplicatestyle="allow")
         pofile.removeduplicates("msgid_comment")
```
I think it is better to instead do the following:

```python
        pofile = self.poparse(posource, duplicatestyle="msgid_comment")
```

unho commented on translate/storage/test_cpo.py:
```diff
@@ -139,8 +136,7 @@ def test_msgid_comment(self):
-        pofile = self.poparse(posource)
-        assert len(pofile.units) == 2
+        pofile = self.poparse(posource, duplicatestyle="allow")
         pofile.removeduplicates("keep")
```
I think it is better to instead do the following:

```python
        pofile = self.poparse(posource, duplicatestyle="keep")
```

unho replied:
What is the difference between `keep` and `allow`?

unho commented on translate/storage/pypo.py:
```diff
@@ -770,6 +773,10 @@ def parse(self, input):
+        if duplicatestyle != "allow":
+            self.removeduplicates(duplicatestyle=duplicatestyle)
+
     def removeduplicates(self, duplicatestyle="merge"):
```
Can you please expand the docstring to list all the acceptable `duplicatestyle` values?

unho commented on translate/storage/test_pypo.py:
```diff
@@ -209,11 +209,32 @@ def test_combine_msgidcomments(self):
+        or converting to other formats
+        """
+        posource = ('#: source1\nmsgid "test me"\nmsgstr ""\n\n'
+                    '#: source2\nmsgid "test me"\nmsgstr ""\n')
```
Please use the following instead which is much more easy to read:

```python
        posource = """#: source1
msgid "test me"
msgstr ""


#: source2
msgid "test me"
msgstr ""
"""
```

Please follow the same pattern in the newly added `posource`.
